File: api/src/main/java/com/prep/api/model/pessoa/Pessoa.java
package com.prep.api.model.pessoa;

import com.prep.api.model.ModelAbstract;
import com.prep.api.model.aluno.Aluno;
import com.prep.api.model.contato.Contato;
import com.prep.api.model.imagem.Imagem;
import com.prep.api.model.interesse.Interesse;
import com.prep.api.model.telefone.Telefone;
import com.prep.api.model.visita.Visita;

import java.util.Date;
import java.util.Set;

public class Pessoa extends ModelAbstract {

    protected Long id;
    protected String nome,
            status,
            localDoCadastro,
            ocupacao;
    protected Character sexo;
    protected Integer idade;
    protected Date dataDoCadastro;

    protected Pessoa responsavel;
    protected Aluno aluno;
    protected Imagem imagem;


    protected Set<? extends Telefone> telefones;
    protected Set<? extends Interesse> interesses;
    protected Set<? extends Contato> contatos;
    protected Set<? extends Visita> visitas;


    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public String getLocalDoCadastro() {
        return localDoCadastro;
    }

    public void setLocalDoCadastro(String localDoCadastro) {
        this.localDoCadastro = localDoCadastro;
    }

    public Character getSexo() {
        return sexo;
    }

    public void setSexo(Character sexo) {
        this.sexo = sexo;
    }

    public Integer getIdade() {
        return idade;
    }

    public void setIdade(Integer idade) {
        this.idade = idade;
    }

    public Date getDataDoCadastro() {
        return dataDoCadastro;
    }

    public void setDataDoCadastro(Date dataDoCadastro) {
        this.dataDoCadastro = dataDoCadastro;
    }

    public String getOcupacao() {
        return ocupacao;
    }

    public void setOcupacao(String ocupacao) {
        this.ocupacao = ocupacao;
    }
}

File: ui/src/app/modules/pessoa/models/pessoa.ts
import { Imagem } from './imagem';
import { Telefone } from './telefone';
import { Aluno } from '../../aluno/models/aluno';
import { Interesse } from '../../interesse/models/interesse';

export class Pessoa {

  public id: number;
  public nome: string;
  public status: string;
  public localDoCadastro: string;
  public sexo: string;
  public idade: number;
  public dataDoCadastro: Date;
  public responsavel: Pessoa;
  public imagem: Imagem ;
  public telefones: Array<Telefone>;
  public interesses: Array<Interesse>;
  public ocupacao: string;
  public aluno: Aluno;

  public static newPessoa() {
    const p = new Pessoa();
    p.dataDoCadastro = new Date();
    p.status = 'ATIVO';
    p.localDoCadastro = 'PREPARA';
    p.telefones = new Array<Telefone>();
    return p;
  }

  // Obter Primeiro e ultimo nome
  public static getFirstAndLastName(fullName: string): string {
    let realeseName = '';
    if (fullName) {
      const names = fullName.split(' ');
      realeseName = this.getFirstName(names) + ' ' + this.getLastName(names);
    }
    return realeseName;
  }

  private static getFirstName(names: Array<string>): string {
    return (names[0]) ? names[0] : '';
  }

  private static  getLastName(names: Array<string>): string {
    return (names.length > 1 && names[names.length - 1]) ?
      names[names.length - 1] : '';
  }

  public isValid(): boolean {
    return this.isNomeValid()
      && this.isStatusValid()
      && this.isDataDoCadastroValid()
      && this.isLocalDoCadastroValid();
  }

  private isNomeValid(): boolean {
    return this.isStringValid(this.nome);
  }

  private isStatusValid(): boolean {
    return this.isStringValid(this.status);
  }

  private isLocalDoCadastroValid(): boolean {
    return this.isStringValid(this.localDoCadastro);
  }

  private isDataDoCadastroValid(): boolean {
    return this.dataDoCadastro !== null;
  }

  private isStringValid(text: string): boolean {
    return (text && text.length !== 0);
  }

}
